backend/testfiles/execution/language/ematch.dark

```diff
@@ -645,4 +644,50 @@ module TypeErrors =
 
   // enum with a single wildcard
   (match TestType.TwoArgs(1, 2) with
-   | TwoArgs _ -> "wildcards allowed") = "wildcards allowed"
\ No newline at end of file
+   | TwoArgs _ -> "wildcards allowed") = "wildcards allowed"
+
+
+module GuardClause =
+  (match 5 with
+   | x when x > 0 -> true
+   | _ -> false) = true
+
+  (match -5 with
+   | x when x > 0 -> true
+   | _ -> false) = false
+
+  (match 6 with
+   | x when x == 5 -> "fail"
+   | -6 -> "fail"
+   | 6 -> "pass"
+   | var -> "fail") = "pass"
+
+  (match 5 with
+   | x when x > 0 -> "positive"
+   | x when x < 0 -> "negative"
+   | _ -> false) = "positive"
+
+  (match (5, "hello") with
+   | (x, y) when x > 0 && y == "hello" -> "pass"
+   | _ -> "fail") = "pass"
+
+  (match (5, "hello") with
+   | (x, y) when x > 0 && y == "hello" -> "first matching branch"
+   | (x, y) when x == 5 && y == "hello" -> "second matching branch"
+   | _ -> "fail") = "first matching branch"
+
+  (match (-5, "hello") with
+   | (x, y) when x > 0 && y == "hello" -> "pass"
+   | _ -> "fail") = "fail"
+
+  (match (5, "hello") with
+   | (x, y) when x > 0 && PACKAGE.Darklang.Stdlib.String.length y == 5 -> "pass"
+   | _ -> "fail") = "pass"
+
+  (match -5 with
+   | x when y > 0 -> true
+   | _ -> false) = Builtin.Test.derrorMessage "There is no variable named: y"
+
+  (match 5 with
```

Comment:
These are great tests. Can you also check what happens if the result of whenCondition is non-boolean?